packages/lit-html/src/lit-html.ts

```diff
@@ -751,7 +784,7 @@ export class NodePart {
     // TODO (justinfagnani): To support nested directives, we'd need to
     // resolve the directive result's values. We may want to offer another
     // way of composing directives.
-    this._setValue(this._directive.update(this, value.values));
+    this._setValue(this._directive._resolve(this, value.values));
```

Comment:
How much benefit do you see out of reusing the part classes in their entirety? If we moved some of the methods to functions, could we use them to build SSR-specific implementations that call directive.render()?

Reply:
So, in SSR we only construct `AttributePart` (the code to run the directive for `NodePart` is chill enough that I just copied it). For `AttributePart`, we basically want to reuse `_resolveValue`, which sequentially resolves each directive, deals with the nothing/noChange, and then concatenates the new string. Even if we made that into a function, we'd still need to thread a flag or function through to do something different on the server. Not sure that's worth it?